docs/architecture/adr-013-metrics.md

```diff
@@ -0,0 +1,90 @@
+# ADR 013: Observability
+
+## Changelog
+
+- 20-01-2020: Initial Draft
+
+## Status
+
+Proposed
+
+## Context
+
+There has been discussion around exposing more metrics to users and node operators about the application. Currently there is only a way to expose metrics from Tendermint and not the application itself. To bring more visibility into applications, I would like to propose reporting of metrics through [Prometheus](https://prometheus.io/).
+
+Extending `AppModuleBasic` to support registering of metrics would enable developers to see more information about individual modules.
+
+```go
+type AppModuleBasic interface {
+  Name() string
+  RegisterCodec(*codec.Codec)
+  RegisterMetrics(namespace string, labelsAndValues ...string) *Metrics
+
+  // genesis
+  DefaultGenesis() json.RawMessage
+  ValidateGenesis(json.RawMessage) error
+
+  // client functionality
+  RegisterRESTRoutes(context.CLIContext, *mux.Router)
+  GetTxCmd(*codec.Codec) *cobra.Command
+  GetQueryCmd(*codec.Codec) *cobra.Command
+}
+// .....
+
+func (bm BasicManager) RegisterMetrics(appName) MetricsProvider {
+	for _, b := range bm {
+		b.RegisterMetrics(appName, labelsAndValues)
+	}
+}
+```
+
+Each module can define its own `CreateMetrics` function:
+
+```go
+type Metrics struct {
+  Size metrics.Guage
+
+  Transactions metrics.Counter
+}
+
+func CreateMetrics(namespace string, labelsAndValues ...string) *Metrics {
+  labels := make([]string, len(labelsAndValues/2))
+  for i := 0; i < len(labelsAndValues); i += 2 {
+      labels[i/2] = labelsAndValues[i]
```

Comment:
I think it's better to define the `labelNames` explicitly as params

Reply:
you mean CreateMetrics(namespace string, labelNames...string, labelValues ...string)?
It is much easier to iterate if it's one array, alternatively we can do map[string]string too which probably is more safe.

Reply:
Actually it's the prometheus API that takes labelsAndValues list as defined above, so if we change it to two lists or even a map, then every module writer needs to convert it themselves. We could provide a helper method but I feel like it's a bit overkill.
@fedekunze are you fine keeping this? you can look at TM example here https://github.com/tendermint/tendermint/blob/c9ef824ddf623a873ca226b98ddb3539c1de1ebc/p2p/metrics.go#L44